### DISCUSSION TITLE
Print every function called in a binary excluding some functions

### DISCUSSION THREAD

**Discussion by h49nakxs** [bpftrace/bpftrace#3597] [Q&A]
Hi there,

I'm trying to print every function that is called in a specific binary excluding some functions.

I've tried the following :
`sudo bpftrace -e 'uprobe:/bin/bash:* / func != "malloc && func != free" / { printf("%s\n", func); }' `

But that doesn't work because of a type mismatch ("usym" and "string").

I'm sure this is possible by writing raw BPF code but do you know if I can achieve that using an existing tool ?

**Comment by ajor** [comment]
I think something like this might work:
```
config = {symbol_source = symbol_table}
uprobe:/bin/bash:* / reg("ip") != uaddr("main") / { printf("%s\n", func); }
```

You'll need the config option, because otherwise bpftrace is free to attach probes to slightly different offsets, e.g. `main+4` instead of `main+0`, to skip the function prologue.

Then we just compare the instruction pointer against the address of the functions you want to exlcude.

**Reply by h49nakxs** [reply]
Hey, that was a quick reply, thanks !

Some notes :

- Unfortunately the filter doesn't work. I tried to print only "main" (result : it doesn't print anything) or print everything but "main" (result : it prints everything) when running a binary compiled with debug info. I've tried with both symbol_source = "dwarf" or "symbol_table".
- symbol_source is not present in the latest release so a build was required (but you probably know it).
- `config = {symbol_source = symbol_table}` triggers an error : **Assignment for symbol_source must be literal**, so I set it with the environment variable BPF_SYMBOL_SOURCE (but I'm not really sure it has been applied though).
- Trying `reg("rip")` for fun generates an error : **'rip' is not a valid register on this architecture (x86_64)**, which is a bit weird isn't it ?

**Reply by ajor** [reply]
> Hey, that was a quick reply, thanks !

Maybe a little too quick - I didn't test it myself :P 

> * Unfortunately the filter doesn't work. I tried to print only "main" (result : it doesn't print anything) or print everything but "main" (result : it prints everything) when running a binary compiled with debug info.

To help with debugging I changed it to print everything out:
```
config = {symbol_source = "symbol_table"}
uprobe:/bin/bash:main { printf("%p %p\n", reg("ip"), uaddr("main")); }
```

On my machine, I'm seeing output like this:
```
0x5608837e43a0 0x303a0
0x563369ca93a0 0x303a0
0x55e19c2f63a0 0x303a0
0x559780f0d3a0 0x303a0
0x5597e33373a0 0x303a0
0x560d566913a0 0x303a0
```

At a glance, it appears that ASLR is messing things up. `uaddr("main")` is returning the static address of `main()` within the ELF, but when bash is loaded into memory and run, `main()` is located at a random offset so the addresses don't match.

With a non-PIE program (I just created a simple C program to test with), the `reg("ip") != uaddr("main")` filter works correctly.

So there appear to be two options:
- recompile your target program without PIE
- disable ASLR on your system, e.g. with `/proc/sys/kernel/randomize_va_space`

> * `config = {symbol_source = symbol_table}` triggers an error : **Assignment for symbol_source must be literal**, so I set it with the environment variable BPF_SYMBOL_SOURCE (but I'm not really sure it has been applied though).

My bad, looks like it wants it as a string: `symbol_source = "symbol_table"`

> * Trying `reg("rip")` for fun generates an error : **'rip' is not a valid register on this architecture (x86_64)**, which is a bit weird isn't it ?

We just take the names that the Linux kernel gives to the registers - that happens to exclude the size prefix for x86_64. Here's the list: https://github.com/bpftrace/bpftrace/blob/3d39d1e18e7f0c0f28d48c2231004f022ae9536f/src/arch/x86_64.cpp#L12-L34

**Comment by jordalgo** [comment]
It is sad that we can't compare the `func` builtin with a string though :(

**Reply by viktormalik** [reply]
Agreed, we should open an issue as it's quite an important functionality. Alternatively, we could allow `func == uaddr("function")` but the kind of comparison should be enabled in a more elegant way from what @ajor suggested above.

**Reply by ajor** [reply]
The type system is happy with this: `func == usym(uaddr("main"))`
But I'm pretty sure it's not doing what people would expect it to do...

**Reply by jordalgo** [reply]
I'll file an issue and look into it.

**Reply by jordalgo** [reply]
https://github.com/bpftrace/bpftrace/issues/3598
